### DISCUSSION TITLE
CH170 Digital

### DISCUSSION THREAD

**Discussion by motoflo** [Nortank12/deepcool-digital-linux#25] [Ideas]
Following information, I get when trying to connect to CH170 Case:

Vendor ID 13875
Device ID 19
Vendor Name DeepCool
Device Name CH170-DIGITAL

Maybe this helps for implementation

**Comment by Nortank12** [comment]
Yes, it helps to identify the device itself. But since you're the first person here who owns a CH170, I have to ask you to test compatibility with existing devices. That will help resolve the method of communication with your device faster.

The highest chance of compatibility is with [AK400 PRO](https://github.com/Nortank12/deepcool-digital-linux/blob/f87d931e6b27858530cf7fc1656b6501794e2ace/src/main.rs#L147) and [CH Series](https://github.com/Nortank12/deepcool-digital-linux/blob/f87d931e6b27858530cf7fc1656b6501794e2ace/src/main.rs#L180).
Simply replace the number with your device ID and check if your display shows anything. If yes, take a picture or a video so we have a better starting point.

I've added [build instructions](https://github.com/Nortank12/deepcool-digital-linux?tab=readme-ov-file#building-from-source) in case you need them.

**Reply by motoflo** [reply]
I tested all PID associated with VID 13875, but no luck. It seems to be a different syntax, as it can also display RPM.

**Reply by Nortank12** [reply]
If you have some time, you can sniff the packets sent on Windows, so it can be reverse-engineered. It worked on the `AK400 DIGITAL PRO`, [here are the instructions](https://github.com/Nortank12/deepcool-digital-linux/issues/21#issuecomment-2535554199).

If you have further questions, feel free to ask.

**Comment by motoflo** [comment]
![image](https://github.com/user-attachments/assets/fb33dccd-61c5-4a04-ab98-88792b70d6c1)

Following on CPU I decoded already:

Power W
10 68 01 06 23 01 02 00 04   [4W]
10 68 01 06 23 01 02 00 08   [8W]

Usage CPU
10 68 01 06 23 01 02 00 09 00 42 30 00 00 63 [99%]
10 68 01 06 23 01 02 00 04 00 42 1C 00 00 06 [6%]

CPU Frequency
10 68 01 06 23 01 02 00 09 00 42 30 00 00 63 07 54 [1876RPM]
10 68 01 06 23 01 02 00 04 00 42 18 00 00 05 07 04 [1796RPM]

CPU Temp is not clear to me:
10 68 01 06 23 01 03 00 09 00 42 50 00 00 64 07 54 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 F0 16
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00   [52°C]

10 68 01 06 23 01 02 00 03 00 42 28 00 00 06 07 54 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 63 16
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00  [41°C]

**Reply by Nortank12** [reply]
Looks like it is very similar to the LD Series and the AK400 DIGITAL PRO. Here's what I figured out (with decimal values):

<table>
    <tr>
        <th>DATA BYTE</th>
        <th>VALUE</th>
        <th>FUNCTION</th>
    </tr>
    <tr>
        <td>D0</td>
        <td>16</td>
        <td>REPORT ID</td>
    </tr>
    <tr>
        <td>D1</td>
        <td>104</td>
        <td rowspan="4">PREAMBLE BYTES <sup>UNCLARIFIED</sup></td>
    </tr>
    <tr>
        <td>D2</td>
        <td>6</td>
    </tr>
    <tr>
        <td>D3</td>
        <td>35</td>
    </tr>
    <tr>
        <td>D4</td>
        <td>1</td>
    </tr>
    <tr>
        <td>D5</td>
        <td>2-3</td>
        <td><i>?</i></td>
    </tr>
    <tr>
        <td>D6</td>
        <td>0-255</td>
        <td rowspan="2">POWER CONSUMPTION <sup><code>U16</code></sup></td>
    </tr>
    <tr>
        <td>D7</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D8</td>
        <td>0-1</td>
        <td><i>TEMPERATURE UNIT ˚C / ˚F?</i></td>
    </tr>
    <tr>
        <td>D9</td>
        <td>0-255</td>
        <td rowspan="4">TEMPERATURE <sup><code>F32</code></sup></td>
    </tr>
    <tr>
        <td>D10</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D11</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D12</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D13</td>
        <td>0-100</td>
        <td>UTILIZATION</td>
    </tr>
    <tr>
        <td>D14</td>
        <td>0-255</td>
        <td rowspan="2">FAN SPEED <sup><code>U16</code></sup></td>
    </tr>
    <tr>
        <td>D15</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D16-D39</td>
        <td>...</td>
        <td><i>NOT USED?</i></td>
    </tr>
    <tr>
        <td>D40</td>
        <td>0-255</td>
        <td>D1-D15 CHECKSUM <sup><code>U8</code> REMAINDER</sup></td>
    </tr>
    <tr>
        <td>D41</td>
        <td>22</td>
        <td>TERMINATION BYTE</td>
    </tr>
    <tr>
        <td>...</td>
        <td>...</td>
        <td>- NOT USED -</td>
    </tr>
</table>

- I saw that D5 can be 2 or 3 based on your examples. I'm not sure why that is.
- D8 should change to 1 if you display fahrenheit instead of celsius.
- D16-D39 might be used when displaying GPU or PSU statistics. In that case, the checksum will be D1-D39.
- I'm pretty confident about the rest.

I'll write the code later, so you can test it.

**Reply by Nortank12** [reply]
```rust
mod devices;
mod monitor;
mod utils;

use devices::*;
use hidapi::HidApi;
use std::{thread::sleep, time::Duration};

const USE_FAHRENHEIT: bool = false;
const POLLING_RATE: u64 = 750;

fn main() {
    // Connect to your device
    let api = HidApi::new().unwrap();
    let device = api.open(13875, 19).unwrap();

    // Set up system monitoring
    let cpu = monitor::cpu::Cpu::new();
    let gpu = monitor::gpu::Gpu::new();

    // Check if `rapl_max_uj` was read correctly
    assert_ne!(cpu.rapl_max_uj, 0);

    // Data packet
    let mut data: [u8; 64] = [0; 64];
    data[0] = 16;
    data[1] = 104;
    data[2] = 1;
    data[3] = 6;
    data[4] = 35;
    data[5] = 1;
    data[6] = 2;

    loop {
        // Initialize the packet
        let mut status_data = data.clone();

        // Read CPU utilization & energy consumption
        let cpu_instant = cpu.read_instant();
        let cpu_energy = cpu.read_energy();

        // Wait
        sleep(Duration::from_millis(POLLING_RATE));

        // ----- CPU DATA -----
        // Power consumption
        let power = (cpu.get_power(cpu_energy, POLLING_RATE)).to_be_bytes();
        status_data[7] = power[0];
        status_data[8] = power[1];

        // Temperature
        let temp = (cpu.get_temp(USE_FAHRENHEIT) as f32).to_be_bytes();
        status_data[9] = if USE_FAHRENHEIT { 1 } else { 0 };
        status_data[10] = temp[0];
        status_data[11] = temp[1];
        status_data[12] = temp[2];
        status_data[13] = temp[3];

        // Utilization
        status_data[14] = cpu.get_usage(cpu_instant);

        // Frequency (fixed to 3800 MHz)
        let frequency = (3800 as u16).to_be_bytes();
        status_data[15] = frequency[0];
        status_data[16] = frequency[1];

        // Fan speed (fixed to 2750 RPM)
        let fanspeed = (2750 as u16).to_be_bytes();
        status_data[17] = fanspeed[0];
        status_data[18] = fanspeed[1];

        // ----- GPU DATA -----
        // Power consumption (fixed to 150 W)
        let power = (150 as u16).to_be_bytes();
        status_data[19] = power[0];
        status_data[20] = power[1];

        // Temperature
        let temp = (gpu.get_temp(USE_FAHRENHEIT) as f32).to_be_bytes();
        status_data[21] = temp[0];
        status_data[22] = temp[1];
        status_data[23] = temp[2];
        status_data[24] = temp[3];

        // Utilization
        status_data[25] = gpu.get_usage();

        // Frequency (fixed to 400 MHz)
        let frequency = (400 as u16).to_be_bytes();
        status_data[26] = frequency[0];
        status_data[27] = frequency[1];

        // Fan speed (fixed to 850 RPM)
        let fanspeed = (850 as u16).to_be_bytes();
        status_data[28] = fanspeed[0];
        status_data[29] = fanspeed[1];

        // ----- PSU DATA -----
        // Power consumption (fixed to 500 W)
        let power = (500 as u16).to_be_bytes();
        status_data[30] = power[0];
        status_data[31] = power[1];

        // Temperature (fixed to 35 C)
        let temp = (35 as f32).to_be_bytes();
        status_data[32] = temp[0];
        status_data[33] = temp[1];
        status_data[34] = temp[2];
        status_data[35] = temp[3];

        // Utilization (fixed to 20)
        status_data[36] = 20;

        // Fan speed (fixed to 600 RPM)
        let fanspeed = (600 as u16).to_be_bytes();
        status_data[37] = fanspeed[0];
        status_data[38] = fanspeed[1];

        // Checksum & termination byte
        let checksum: u16 = status_data[1..=39].iter().map(|&x| x as u16).sum();
        status_data[40] = (checksum % 256) as u8;
        status_data[41] = 22;

        device.write(&status_data).unwrap();
    }
}
```

Replace `main.rs`with this code.

**Reply by motoflo** [reply]
I tried and it is recognizing the device but does not display anything. I get quite some build warnings.

**Reply by Nortank12** [reply]
Build warnings should be fine since you're running only what is necessary, making most of the code considered dead code.

I've made a mistake by skipping over D2. I updated the code and added the values for GPU and PSU statistics. It should be working now. If not, try to check if I'm still missing a value somewhere and make sure you run it as root (if you haven't set up the udev rules).

**Reply by motoflo** [reply]
Yes it works now and depending on D6 following is displayed:

D6=1 Like display test enable all fields in  the display
D6=2 shows CPU_Temp CPU_Power CPU_Usage and CPU_Frequency
D6=3 shows CPU_Temp CPU_Power CPU_Usage CPU_ Fan_RPM
D6=4 Shows GPU_Temp GPU_Power GPU _Usage GPU_Freq
D6=5 Shows PSU_Temp PSU_Power PSU_Usage PSU_Fan_RPM
D6=6  Shows PSU_Temp PSU_Power PSU_Usage PSU_??W

**Reply by motoflo** [reply]
There seem to be some mapping issues as it shows 850 for PSU Power as this seem to be Fan speed

        // Fan speed (fixed to 850 RPM)
        let fanspeed = (850 as u16).to_be_bytes();
        status_data[28] = fanspeed[0];
        status_data[29] = fanspeed[1];

        // ----- PSU DATA -----
        // Power consumption (fixed to 500 W)
        let power = (500 as u16).to_be_bytes();
        status_data[30] = power[0];
        status_data[31] = power[1];

**Reply by Nortank12** [reply]
Looks like we're on a good track. Just to clarify:
- Will the PSU power offset the rest of the values? Meaning PSU temperature will be [30, 31, 32, 33] and so on...
- Did you try to modify D9 to see if it changes the temperature unit to Fahrenheit?
- Does changing the value of D39 make any difference?

**Comment by motoflo** [comment]
I finished testing of the other modes. In CPU Mode it swaps constantly between displaying FAN and Freq and I think this is what D5 is for. I tried to catch it in the USB log.

Mode selection CPU

Displaying RPM

10 68 01 06 23 01 03 00 38 00 42 44 00 00 05 12 B4 03 DF 00 00 42 2C 00 00 00 00 D2 00
00 00 00 00 00 00 00 00 00 00 00 41 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00  [984RPM]

Displaying Freq

10 68 01 06 23 01 02 00 1D 00 42 50 00 00 05 13 18 03 D6 00 00 42 1C 00 00 0F 00 D2 00
00 00 00 00 00 00 00 00 00 00 00 8C 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 [4888MHz]

PSU Mode is all 0 seems not compatible with my setup:

Mode selection PSU

10 68 01 06 23 01 04 00 2B 00 42 38 00 00 05 13 18 03 BC 00 00 42 A2 00 00 64 05 CD 00
00 00 00 00 00 00 00 00 00 00 00 45 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00  [0°] [0W] [0%] [0RPM]

GPU selection in software:

Mode selection GPU

10 68 01 06 23 01 04 00 16 00 42 44 00 00 03 13 18 03 DF 00 00 42 48 00 00 04 00 D2 00
00 00 00 00 00 00 00 00 00 00 00 A3 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 [50°] [0%] [0W] [210MHz]

10 68 01 06 23 01 04 00 23 00 42 4C 00 00 06 13 19 03 93 00 00 42 A6 00 00 64 05 CD 00
00 00 00 00 00 00 00 00 00 00 00 2E 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 [82°] [100%] [0W] [1485MHz]

**Reply by Nortank12** [reply]
Here's a modified mapping table that might be closer to how it works:

<table>
    <tr>
        <th>DATA BYTE</th>
        <th>VALUE</th>
        <th>FUNCTION</th>
        <th>DISPLAY</th>
    </tr>
    <tr>
        <td>D0</td>
        <td>16</td>
        <td>REPORT ID</td>
        <td rowspan="7">-</td>
    </tr>
    <tr>
        <td>D1</td>
        <td>104</td>
        <td rowspan="5">PREAMBLE BYTES <sup>UNCLARIFIED</sup></td>
    </tr>
    <tr>
        <td>D2</td>
        <td>1</td>
    </tr>
    <tr>
        <td>D3</td>
        <td>6</td>
    </tr>
    <tr>
        <td>D4</td>
        <td>35</td>
    </tr>
    <tr>
        <td>D5</td>
        <td>1</td>
    </tr>
    <tr>
        <td>D6</td>
        <td>2-4</td>
        <td><i>DISPLAY MODE?</i></td>
    </tr>
    <tr>
        <td>D7</td>
        <td>0-255</td>
        <td rowspan="2">POWER CONSUMPTION <sup><code>U16</code></sup></td>
        <td rowspan="2">CPU</td>
    </tr>
    <tr>
        <td>D8</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D9</td>
        <td>0-1</td>
        <td><i>TEMPERATURE UNIT ˚C / ˚F?</i></td>
        <td>ALL</td>
    </tr>
    <tr>
        <td>D10</td>
        <td>0-255</td>
        <td rowspan="4">TEMPERATURE <sup><code>F32</code></sup></td>
        <td rowspan="9">CPU</td>
    </tr>
    <tr>
        <td>D11</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D12</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D13</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D14</td>
        <td>0-100</td>
        <td>UTILIZATION</td>
    </tr>
    <tr>
        <td>D15</td>
        <td>0-255</td>
        <td rowspan="2">FREQUENCY <sup><code>U16</code></sup></td>
    </tr>
    <tr>
        <td>D16</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D17</td>
        <td>0-255</td>
        <td rowspan="2">FAN SPEED <sup><code>U16</code></sup></td>
    </tr>
    <tr>
        <td>D18</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D19</td>
        <td>0-255</td>
        <td rowspan="2">POWER CONSUMPTION <sup><code>U16</code></sup></td>
        <td rowspan="11">GPU</td>
    </tr>
    <tr>
        <td>D20</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D21</td>
        <td>0-255</td>
        <td rowspan="4">TEMPERATURE <sup><code>F32</code></sup></td>
    </tr>
    <tr>
        <td>D22</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D23</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D24</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D25</td>
        <td>0-100</td>
        <td>UTILIZATION</td>
    </tr>
    <tr>
        <td>D26</td>
        <td>0-255</td>
        <td rowspan="2">FREQUENCY <sup><code>U16</code></sup></td>
    </tr>
    <tr>
        <td>D27</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D28</td>
        <td>0-255</td>
        <td rowspan="2">FAN SPEED <sup><code>U16</code></sup></td>
    </tr>
    <tr>
        <td>D29</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D30</td>
        <td>0-255</td>
        <td rowspan="2">PSU POWER CONSUMPTION <sup><code>U16</code></sup></td>
        <td rowspan="10">PSU</td>
    </tr>
    <tr>
        <td>D31</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D32</td>
        <td>0-255</td>
        <td rowspan="4"> PSU TEMPERATURE <sup><code>F32</code></sup></td>
    </tr>
    <tr>
        <td>D33</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D34</td>
        <td>0-255</td>
    </tr>
    <tr>
        <td>D35</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D36</td>
        <td>0-100</td>
        <td>PSU UTILIZATION</td>
    </tr>
    <tr>
        <td>D37</td>
        <td>0-255</td>
        <td rowspan="2">PSU FAN SPEED <sup><code>U16</code></sup></td>
    </tr>
    <tr>
        <td>D38</td>
        <td>1-255</td>
    </tr>
    <tr>
        <td>D39</td>
        <td>0</td>
        <td>?</td>
    </tr>
    <tr>
        <td>D40</td>
        <td>0-255</td>
        <td>D1-D39 CHECKSUM <sup><code>U8</code> REMAINDER</sup></td>
        <td rowspan="2">-</td>
    </tr>
    <tr>
        <td>D41</td>
        <td>22</td>
        <td>TERMINATION BYTE</td>
    </tr>
    <tr>
        <td>...</td>
        <td>...</td>
        <td>- NOT USED -</td>
        <td>...</td>
    </tr>
</table>

D6 could be a mode switch [CPU, GPU, PSU] as well.

**Reply by motoflo** [reply]
There seems to be no GPU FAN field so a slight shift in Data set  and possition:

    // PSU Power(fixed to 850 RPM)
        let fanspeed = (855 as u16).to_be_bytes();
        status_data[28] = fanspeed[0];
        status_data[29] = fanspeed[1];

        // ----- PSU DATA -----
        // Power consumption (fixed to 500 W)
        let power = (500 as u16).to_be_bytes();
        status_data[35] = power[0];
        status_data[36] = power[1];

        // Temperature (fixed to 35 C)
        let temp = (35 as f32).to_be_bytes();
        status_data[30] = temp[0];
        status_data[31] = temp[1];
        status_data[32] = temp[2];
        status_data[33] = temp[3];

        // Utilization (fixed to 20)
        status_data[34] = 20;

        // Fan speed (fixed to 600 RPM)
        let fanspeed = (600 as u16).to_be_bytes();
        status_data[37] = fanspeed[0];
        status_data[38] = fanspeed[1];

D9 is also working as a switch between F and C

All should be correct now

**Reply by Nortank12** [reply]
Perfect! I've tweaked the code a bit, so it only reads what will be displayed. CPU data is sent when display mode is CPU, and the same for the GPU. The PSU is more unique, maybe DeepCool will release their own units that will support it, so it displays a warning for now. I've looked into CPU fan speed monitoring on Linux, and there's no universal driver for it, so I had to display a warning for that too.

The changes are in the main branch now, and everything should work by default. Try to test the C/F switch and the 2 modes to see if they work correctly. If yes, I'll make a new release and mark the CH170 DIGITAL as tested.

**Reply by motoflo** [reply]
Tested CPU and GPU in C and F all confirmed working

**Reply by Nortank12** [reply]
I updated the release. Thanks for your contribution!
